### PR TITLE
enhance(unstruct): adds new version of unstruct package

### PR DESCRIPTION
**Pull request by ashishranjan738**
Signed-off-by: Ashish Ranjan <ashish.ranjan@mayadata.io>

**What this PR does / why we need it**:
This commit adds a new version of `unstruct package` which has been written in maya
idiomatic fashion.

**Special notes for your reviewer**:

**Checklist:**
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [x] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: pkg/unstruct/v1alpha2/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+			return err
+		}
+		k.Client = cli
+	}
+	return nil
+}
+
+// WithKubeClient sets the kubernetes client against
+// the kubeclient instance
+func WithKubeClient(c client.Client) kubeclientBuildOption {
```
Comment (AmitKumarDas) [inline pkg/unstruct/v1alpha2/kubernetes.go]: Let us rename this to `WithClient` else it is becoming repetitive with `KubeClient`

File: pkg/unstruct/v1alpha2/unstruct.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+// to API's unstructured
+func (u *unstruct) ToUnstructured() *unstructured.Unstructured {
+	return u.object
+}
+
+// builder enables building of an
+// unstructured instance
+type builder struct {
+	unstruct *unstruct
+	err      error
```
Comment (AmitKumarDas) [inline pkg/unstruct/v1alpha2/unstruct.go]: please make this as `errs []error`

File: pkg/unstruct/v1alpha2/unstruct_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		expectError                bool
+	}{
+		"Test 1": {fakeK8sResource, "icstcee", false},
+		"Test 2": {fakeInvalidK8sResource, "", true},
+	}
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			b := UnstructBuilderForYaml(mock.resourceYAML)
+			if mock.expectError && b.err == nil {
+				t.Fatalf("Test %q failed, expected err but got nil", name)
```
Comment (AmitKumarDas) [inline pkg/unstruct/v1alpha2/unstruct_test.go]: what is the value provided via %q ? Why not %s?

File: pkg/unstruct/v1alpha2/unstruct_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+			} else if b.unstruct.object.GetName() != mock.expectedName {
+				t.Fatalf("Test %q failed, expected %v but got %v", name, mock.expectedName, b.unstruct.object.GetName())
```
Comment (AmitKumarDas) [inline pkg/unstruct/v1alpha2/unstruct_test.go]: what is the value provided via %q ? Why not %s?
we we rephrase this as:
`test '%s' failed: expected name '%s' got '%s'`

File: pkg/unstruct/v1alpha2/unstruct_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	}{
+		"Test 1": {"icstcee", "icstcee"},
+		"Test 2": {"icstcee1", "icstcee1"},
+	}
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			mockObj := fakeUnstructObject(mock.resourceName)
+			b := UnstructBuilderForObject(mockObj)
+			if b.unstruct.object.GetName() != mock.expectedName {
+				t.Fatalf("Test %q failed, expected %v but got %v", name, mock.expectedName, b.unstruct.object.GetName())
```
Comment (AmitKumarDas) [inline pkg/unstruct/v1alpha2/unstruct_test.go]: what is the value provided via %q ? Why not %s?

File: pkg/unstruct/v1alpha2/unstruct_test.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	`
+)
+
+func fakeUnstructObject(name string) *unstructured.Unstructured {
+	u := &unstructured.Unstructured{}
+	u.SetName(name)
+	return u
+}
+
+func TestUnstructForYAMLBuilder(t *testing.T) {
```
Comment (AmitKumarDas) [inline pkg/unstruct/v1alpha2/unstruct_test.go]: Can you rename based on the target function i.e. the function that is being tested here.

File: pkg/unstruct/v1alpha2/unstruct_test.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+			mockObj := fakeUnstructObject(mock.resourceName)
+			b := UnstructBuilderForObject(mockObj)
+			if b.unstruct.object.GetName() != mock.expectedName {
+				t.Fatalf("Test %q failed, expected %v but got %v", name, mock.expectedName, b.unstruct.object.GetName())
+			}
+		})
+	}
+}
+
+func TestUnstructBuild(t *testing.T) {
```
Comment (AmitKumarDas) [inline pkg/unstruct/v1alpha2/unstruct_test.go]: can we rename to `TestUnstructBuilderForYamlBuild`
I understand this is a long name but this reflects the receiver as well the method that is under test

File: pkg/unstruct/v1alpha2/unstruct.go (hunk 2 above)
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/unstruct.go]: done

File: pkg/unstruct/v1alpha2/kubernetes.go (hunk 1 above)
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/kubernetes.go]: done

File: pkg/unstruct/v1alpha2/unstruct_test.go (hunk 6 above)
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/unstruct_test.go]: changed

File: pkg/unstruct/v1alpha2/unstruct_test.go (hunk 3 above)
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/unstruct_test.go]: Since %q will print the name in quotes so I thought it will look good, anyway changed back to %s.

File: pkg/unstruct/v1alpha2/unstruct_test.go (hunk 4 above)
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/unstruct_test.go]: Since %q will print the name in quotes so i thought it will look good, anyway changed back to %s.

File: pkg/unstruct/v1alpha2/unstruct_test.go (hunk 5 above)
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/unstruct_test.go]: Since %q will print the name in quotes so i thought it will look good, anyway changed back to %s.

File: pkg/unstruct/v1alpha2/unstruct_test.go (hunk 7 above)
Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/unstruct_test.go]: done

File: pkg/unstruct/v1alpha2/unstruct.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	b := UnstructBuilder()
+	b.unstruct.object = obj
+	return b
+}
+
+// Build returns the unstruct object created by
+// the builder
+func (b *builder) Build() (*unstruct, error) {
+	if len(b.errs) != 0 {
+		return nil, fmt.Errorf("%v", b.errs)
```
Comment (AmitKumarDas) [inline pkg/unstruct/v1alpha2/unstruct.go]: We shall not use fmt for throwing error
We shall use `github.com/pkg/errors` only

Comment (ashishranjan738) [reply pkg/unstruct/v1alpha2/unstruct.go]: done
